# TC1002S Herramientas computacionales: el arte de la analítica

This is a notebook with all your work for the final evidence of this course


# Niveles de dominio a demostrar con la evidencia

### SING0202A

Interpreta interacciones entre variables relevantes en un problema, como base para la construcción de modelos bivariados basados en datos de un fenómeno investigado que le permita reproducir la respuesta del mismo. Es capaz de construir modelos bivariados que expliquen el comportamiento de un fenómeno.


# Student information

- Name: Ricardo Gutiérrez Juárez

- ID: A01637398

- My carreer: IRS


# Importing libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.cluster import KMeans
from sklearn.datasets import load_digits

# PART 1

## Do clustering using your assigned dataset

## a) Load data

In [2]:
data = pd.read_csv("/Users/rguti/Documents/TC1002s/TC1002S/NotebooksStudents/A01637398/A01637398_X.csv")

## b) Data managment


Print the first 7 rows

In [3]:
print(data.head(7))

   Unnamed: 0        x1        x2         x3         x4        x5        x6  \
0           0  5.915624  3.619747 -11.374853   4.511048  8.473472 -0.361552   
1           1  9.163681  3.243405  -9.117147   4.914047  8.672830  3.087021   
2           2 -8.811333 -7.833458   8.160205   6.704484 -5.916631  3.805485   
3           3 -5.637697  0.470401  -2.524303   1.921583  7.541389  7.160939   
4           4 -5.671844  2.842315  -7.056645   5.438295  3.549798  5.525150   
5           5  6.301221  6.730163   5.753297 -10.307334  1.088196 -0.480127   
6           6 -7.431247 -8.087101   7.501752   9.577647 -8.835624  1.932257   

         x7         x8        x9       x10  
0  4.870508   5.611644  6.442019 -6.215415  
1  5.912965   8.436283  8.906484 -9.544977  
2 -4.482368  -7.269878 -4.727307  5.715858  
3  1.345485   4.397946  6.049104  3.362854  
4  6.206323   4.782128  8.909895  3.783253  
5  0.916543   1.885479 -0.002643  8.136968  
6 -1.994726 -11.371362 -1.454350  3.099336  


Print the last 4 rows

In [4]:
print(data.tail(4))

     Unnamed: 0        x1        x2        x3         x4        x5        x6  \
731         731 -5.637308 -1.260777 -0.465104   7.413143  5.308432  6.697013   
732         732 -4.718331  0.428782 -4.248313   4.501124  6.734380  6.284131   
733         733 -7.073926 -5.567657  5.123849  10.245982 -8.480121 -3.590681   
734         734  4.468632  3.716021  6.367471  -7.726538  7.122860  1.986626   

           x7        x8        x9       x10  
731  2.785377  5.754246  6.186965  1.561866  
732  4.360696  4.882850  6.866648  3.708612  
733 -3.264721 -9.229966 -1.771476  4.022909  
734  2.893201  2.164670 -3.859932  6.392884  


How many rows and columns are in your data?

    There are 734 rows

Use the `shape` method


In [ ]:
print("Shape:", data.shape)

Print the name of all columns

Use the `columns` method

In [5]:
print("Columns:", data.columns)

Columns: Index(['Unnamed: 0', 'x1', 'x2', 'x3', 'x4', 'x5', 'x6', 'x7', 'x8', 'x9',
       'x10'],
      dtype='object')


What is the data type in each column

    They are float type

Use the `dtypes` method

In [6]:
print("Data Types:\n", data.dtypes)

Data Types:
 Unnamed: 0      int64
x1            float64
x2            float64
x3            float64
x4            float64
x5            float64
x6            float64
x7            float64
x8            float64
x9            float64
x10           float64
dtype: object


What is the meaning of rows and columns?

    The rows represent individual observations, and each column represents a variable or feature. Columns may contain different types of data, and these variables will be analyzed during clustering.

Print a statistical summary of your columns


In [9]:
data.describe()

Unnamed: 0          x1          x2          x3          x4          x5  \
count  735.000000  735.000000  735.000000  735.000000  735.000000  735.000000   
mean   367.000000   -0.520510   -0.170266    0.534681    3.380816    3.128626   
std    212.320512    6.423091    6.416149    7.465832    7.553236    7.203315   
min      0.000000  -14.650917  -14.671289  -13.191483  -14.479417  -13.739735   
25%    183.500000   -6.479684   -5.078483   -6.348451   -1.354786   -1.704289   
50%    367.000000   -1.258502    1.264727    1.499409    6.510904    5.647765   
75%    550.500000    5.523403    4.968451    7.499789    8.477703    8.272257   
max    734.000000   11.645005   12.047021   13.755580   14.510592   15.065879   

               x6          x7          x8          x9         x10  
count  735.000000  735.000000  735.000000  735.000000  735.000000  
mean     2.226895    1.035172    1.099545    1.616662    2.394816  
std      3.044472    3.762569    6.633118    6.009506    6.178402  
min     -4.604168  -10.624645  -13.931141   -8.858669  -12.612024  
25%      0.000697   -1.572476   -3.562742   -3.859282   -1.344017  
50%      1.817137    1.955278    3.334973    0.779197    4.361926  
75%      4.341546    3.667978    5.611232    6.942227    6.863970  
max     12.818733    9.208096   12.311866   15.016982   12.546748

1. What is the minumum and maximum values of each variable:

min	x0:0.000000	x1:-14.650917	x2:-14.671289	x3:-13.191483	x4:-14.479417	x5:-13.739735	x6-4.604168	x7:-10.624645	x8:-13.931141	x9:-8.858669	x10:-12.612024

max	734.000000	11.645005	12.047021	13.755580	14.510592	15.065879	12.818733	9.208096	12.311866	15.016982	12.546748

2. What is the mean and standar deviation of each variable:

mean	367.000000	-0.520510	-0.170266	0.534681	3.380816	3.128626	2.226895	1.035172	1.099545	1.616662	2.394816

3. What the 25%, 50% and 75% represent?:

25%	183.500000	-6.479684	-5.078483	-6.348451	-1.354786	-1.704289	0.000697	-1.572476	-3.562742	-3.859282	-1.344017

50%	367.000000	-1.258502	1.264727	1.499409	6.510904	5.647765	1.817137	1.955278	3.334973	0.779197	4.361926

75%	550.500000	5.523403	4.968451	7.499789	8.477703	8.272257	4.341546	3.667978	5.611232	6.942227	6.863970

Rename the columns using the same name with capital letters

In [10]:
data.columns = [col.upper() for col in data.columns]
print(data.head())

   UNNAMED: 0        X1        X2         X3        X4        X5        X6  \
0           0  5.915624  3.619747 -11.374853  4.511048  8.473472 -0.361552   
1           1  9.163681  3.243405  -9.117147  4.914047  8.672830  3.087021   
2           2 -8.811333 -7.833458   8.160205  6.704484 -5.916631  3.805485   
3           3 -5.637697  0.470401  -2.524303  1.921583  7.541389  7.160939   
4           4 -5.671844  2.842315  -7.056645  5.438295  3.549798  5.525150   

         X7        X8        X9       X10  
0  4.870508  5.611644  6.442019 -6.215415  
1  5.912965  8.436283  8.906484 -9.544977  
2 -4.482368 -7.269878 -4.727307  5.715858  
3  1.345485  4.397946  6.049104  3.362854  
4  6.206323  4.782128  8.909895  3.783253  


Rename the columns to their original names

In [12]:
data.columns = [col.lower() for col in data.columns]
print(data.head())

   unnamed: 0        x1        x2         x3        x4        x5        x6  \
0           0  5.915624  3.619747 -11.374853  4.511048  8.473472 -0.361552   
1           1  9.163681  3.243405  -9.117147  4.914047  8.672830  3.087021   
2           2 -8.811333 -7.833458   8.160205  6.704484 -5.916631  3.805485   
3           3 -5.637697  0.470401  -2.524303  1.921583  7.541389  7.160939   
4           4 -5.671844  2.842315  -7.056645  5.438295  3.549798  5.525150   

         x7        x8        x9       x10  
0  4.870508  5.611644  6.442019 -6.215415  
1  5.912965  8.436283  8.906484 -9.544977  
2 -4.482368 -7.269878 -4.727307  5.715858  
3  1.345485  4.397946  6.049104  3.362854  
4  6.206323  4.782128  8.909895  3.783253  


Use two different alternatives to get one of the columns

In [15]:
column_data = data['x1']
print(column_data)

column_data = data.x1
print(column_data)

0      5.915624
1      9.163681
2     -8.811333
3     -5.637697
4     -5.671844
         ...   
730   -6.490188
731   -5.637308
732   -4.718331
733   -7.073926
734    4.468632
Name: x1, Length: 735, dtype: float64
0      5.915624
1      9.163681
2     -8.811333
3     -5.637697
4     -5.671844
         ...   
730   -6.490188
731   -5.637308
732   -4.718331
733   -7.073926
734    4.468632
Name: x1, Length: 735, dtype: float64


Get a slice of your data set: second and thrid columns and rows from 62 to 72

In [17]:
data_slice = data.iloc[62:73, 2:4] 
print(data_slice)

           x2         x3
62  -3.845272  -5.896392
63  -2.589709  -8.521298
64   0.524676  -4.448797
65   1.814764 -10.893499
66   7.264946  -7.657977
67   3.824875 -11.541880
68   1.080268  -4.600395
69   3.481305  -7.879416
70   7.650758   8.983825
71   4.071828   8.541432
72 -11.720737   6.545526


For the second and thrid columns, calculate the number of null and not null values and verify that their sum equals the total number of rows

In [20]:
second_column = data.iloc[:, 2]
third_column = data.iloc[:, 3]

second_column_nulls = second_column.isnull().sum()
second_column_not_nulls = second_column.notnull().sum()

third_column_nulls = third_column.isnull().sum()
third_column_not_nulls = third_column.notnull().sum()

print(f"Second Column - Null values: {second_column_nulls}, Not Null values: {second_column_not_nulls}")
print(f"Third Column - Null values: {third_column_nulls}, Not Null values: {third_column_not_nulls}")

total_rows = len(data)
print(f"Total Rows: {total_rows}")
print(f"Second Column - Null + Not Null: {second_column_nulls + second_column_not_nulls == total_rows}")
print(f"Third Column - Null + Not Null: {third_column_nulls + third_column_not_nulls == total_rows}")

Second Column - Null values: 0, Not Null values: 735
Third Column - Null values: 0, Not Null values: 735
Total Rows: 735
Second Column - Null + Not Null: True
Third Column - Null + Not Null: True


Discard the last column


### Questions

Based on the previos results, provide a full description of yout dataset

Your response:




## c) Data visualization

Plot in the histogram of one of the variables

Plot in the same figure the histogram of two variables



Based on these plots, provide a description of your data:

Your response here:

Plot the boxplot of one of the variables

Plot in the same figure the boxplot of two variables

Based on these plots, provide a description of your data:

Your response here:

Plot the scatter plot between all pair of variables

### Questions

Based on the previos plots, provide a full description of yout dataset

Your response:




## d) Kmeans

Do Kmeans clustering assuming a number of clusters accorging to your scatter plots

Add to your dataset a column with the estimated cluster to each data point

Print the number associated to each cluster

Print the centroids

Print the intertia metric

Plot a scatter plot of your data using different color for each cluster. Also plot the centroids

### Questions

Provides a detailed description of your results

Your response:

## d) Elbow plot

Compute the Elbow plot

### Questions

What is the best number of clusters K? (argue your response)

Your response:


Does this number of clusters agree with your inital guess? (argue your response, no problem at all if they do not agree)

Your response:



# PART 2

## Do clustering using the "digits" dataset

1) Load the dataset from "sklearn.datasets"


2) Plot some of the observations (add in the title the label/digit of that obserbation)

3) Do K means clustering in the following cases:

* KmeansAll: Using all 64 variables/pixels/features

* Kmeans1row: Using only the 8 variables/pixels/features from the firt row

* Kmeans4row: Using only the 8 variables/pixels/features from the fourth row

* Kmeans8row: Using only the 8 variables/pixels/
features from the eighth row


4) Verify your results. Plot several observations from the same digit and add in the title the real label and the estimated label to check in what observations the clusterization was correct or incorrect

5) Compute the Elbow plot

### Questions

Provides a detailed description of your results (e.g., in which case the clusterization is better, with KmeansAll, Kmeans1row, Kmeans4row, or Kmeans8row).

Your response (argue your response):

# PART 3
## Do classification using the "digits" dataset

1) Load the dataset from "sklearn.datasets"

2) Plot some of the observations  (add in the title the label/digit of that obserbation)

3) Split the dataset in train and test

4) Tune a classifier (Use the train set) in the following cases:

* ClassifierAll: Using all 64 variables/pixels/features

* Classifier1col: Using only the 8 variables/pixels/features from the firt column

* Classifier4col: Using only the 8 variables/pixels/features from the fourth column

* Classifier8col: Using only the 8 variables/pixels/
features from the eighth column

Note: in these four cases always use the same classification algorithm, e.g., a Suport Vector Machine


5) Make predictions (use the test set)

6) Compute performance metrics

7) Verify your results. Plot several observations from the same digit and add in the title the real label and the estimated label to check in what observations the classification was correct or incorrect

### Questions

Provides a detailed description of your results (e.g., in which case the classification performance is better, with ClassifierAll, Classifier1col, Classifier4col, or Classifier8col).

Your response (argue your response):

# PART 4

## Descripción de tu percepcion del nivel de desarrollo de la subcompetencia

### SING0202A Interpretación de variables


Escribe tu description del nivel de logro del siguiente criterio de la subcompetencia

**Interpreta interacciones**. Interpreta interacciones entre variables relevantes en un problema, como base para la construcción de modelos bivariados basados en datos de un fenómeno investigado que le permita reproducir la respuesta del mismo.

Tu respuesta:

Escribe tu description del nivel de logro del siguiente criterio de la subcompetencia

**Construcción de modelos**. Es capaz de construir modelos bivariados que expliquen el comportamiento de un fenómeno.

Tu respuesta: